[Wesley Teo] iP

Bisceto opened the pull request:
# Duke Chatbot
> It is only when one goes through CS2103T, does a man truly know the value of time.

With Duke Chatbot, you'll be able to easily track your tasks and events! Have more time for yourself and for CS2106!

Duke is: 
- A text-based management system
- Easy to learn
- Intuitive to use 


To get started: 
1. Download the latest release from [here](https://github.com/Bisceto/ip/releases/tag/A-Jar).
2. Navigate to the folder using a terminal
3. Run `java -jar Duke.jar` (Ensure you have java installed on your computer!)
4. Get crazy and get productive! 🥇 

Features:

 * [x]Managing tasks
 * [x]Managing deadlines
 * [ ]Reminders 
If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```
public static void main(String[] args) {
        new Duke("../../../data/duke.txt").run();
  }
```

Commit:
Add Gradle support

Bisceto committed:
Added inputs for Duke and made Duke repeat the input.

Bisceto committed:
Added list functionality and adding items to list

Bisceto committed:
Added mark feature to mark tasks as done or undone

Bisceto committed:
Added categories feature to tasks. Added Event.java, Todo.java, Deadline.java. Separated queries in to Query.java

Bisceto committed:
Added automated text UI testing

Bisceto committed:
Added error handling features when there is a missing description for tasks or invalid commands

Bisceto committed:
Added delete task feature

Bisceto committed:
Added save functionality to duke.txt

Bisceto committed:
Added functionality to understand dates and times.

Bisceto committed:
Refactored code to include more OOP

Bisceto committed:
Organised classes into packages

Bisceto committed:
Merge branch 'add-gradle-support' of https://github.com/Bisceto/ip

Bisceto committed:
Added gradle support

SeanChinJunKai commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,58 @@
+
+import duke.task.*;
+
+import java.io.*;
```
Please list imported classes explicitly.

SeanChinJunKai commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,42 @@
+
+    public void run() {
+        ui.showWelcomeMessage();
+        boolean isExit = false;
```
The boolean variable isExit explains the named entity to the reader accurately and at a sufficient level of detail.

SeanChinJunKai commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,50 @@
+    protected LocalDateTime startTime;
+    protected LocalDateTime endTime;
+    protected DateTimeFormatter inputFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+    protected DateTimeFormatter displayFormatter = DateTimeFormatter.ofPattern("MMM dd yyyy HHmm");
```
Would it be good to leave an empty line between the variables and methods?

SeanChinJunKai commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,24 @@
+
+    public Todo(String input, boolean isDone) {
+        super(input, isDone);
+        this.symbol = 'T';
```
Would it be better to include the symbol 'T' inside the super()?

SeanChinJunKai commented on src/main/java/duke/Query.java:
```diff
@@ -0,0 +1,47 @@
+            return QueryType.exit;
+        }
+        String[] inputArr = input.split(" ");
+        if (inputArr.length == 2 && isNumeric(inputArr[1])) {
```
Job well done on avoiding arrowhead style of code.

SeanChinJunKai reviewed (commented):
LGTM. Just some nits to fix.

nus-jeff commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,62 @@
+        case delete:
+            return new DeleteCommand(Integer.parseInt(fullCommandArr[1]));
+        case exit:
+            return new ExitCommand();
```
Maybe you would like to add a break statement or fallthrough comment after each case to improve code reading?

nus-jeff commented on src/main/java/duke/Query.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    static boolean isNumeric(String str) {
+        char[] charArray = str.toCharArray();
```
Nice that you placed the array sign with the variable type 👍 , perhaps you could use a plural variable name here to represent collection of chars?

nus-jeff commented on src/main/java/duke/Query.java:
```diff
@@ -0,0 +1,47 @@
+        } else if (input.equals("bye")) {
+            return QueryType.exit;
+        }
+        String[] inputArr = input.split(" ");
```
Perhaps you could use a plural variable name here to represent collection of String?

nus-jeff commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,58 @@
+public class Storage {
+    protected File file;
+
+    public Storage(String fileInput) throws DukeException{
```
Probably you have missed this out, but maybe you would want to have a space between DukeException and the bracket.
```suggestion
   public Storage(String fileInput) throws DukeException {
```

nus-jeff commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,58 @@
+                Files.createDirectories(Paths.get(file.getParent()));
+                this.file = Files.createFile(Path.of(fileInput)).toFile();
+            }
+        } catch (IOException err){
```
Probably you have missed this out, but maybe you would want to have a space between the catch exception and the bracket.

```suggestion
 } catch (IOException err) {

nus-jeff commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,26 @@
+import duke.task.TaskList;
+
+public class DeleteCommand extends Command {
+    private final int index;
```
Maybe you would like to use full caps to represent index, since its a constant?

nus-jeff reviewed (commented):
Overall really good code, super easy to follow and understand 👍 
Only minor formatting may be needed and your code becomes even better than before!

Bisceto committed:
Added JUnit test cases for testing.

Bisceto committed:
Added JavaDoc comments to public methods

Bisceto committed:
Merge branch 'branch-A-JavaDoc'

Bisceto committed:
Edited code to follow coding standard

Bisceto committed:
Merge branch 'branch-A-JavaDoc'

Bisceto committed:
Edited code to follow coding standards.

Bisceto committed:
Merge branch 'branch-A-CodingStandard'

Bisceto committed:
Worked on Find Command

Bisceto committed:
Added Find Tasks functionality

Bisceto committed:
Added Launcher.java

pangrwa commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,83 @@
+        QueryType inputType = Query.queryType(fullCommand);
+        String[] fullCommandArr = fullCommand.split(" ");
+
+        switch (inputType) {
+        case list:
+            return new ListCommand();
+        case todo:
+            Todo todo = new Todo(fullCommand);
+            return new AddCommand(todo);
+        case deadline:
+            Deadline deadline = new Deadline(fullCommand);
+            return new AddCommand(deadline);
+        case event:
+            Event event = new Event(fullCommand);
+            return new AddCommand(event);
+        case find:
+            return new FindCommand(fullCommandArr[1]);
+        case mark:
+            return new MarkCommand(Integer.parseInt(fullCommandArr[1]));
+        case unmark:
+            return new UnmarkCommand(Integer.parseInt(fullCommandArr[1]));
+        case delete:
+            return new DeleteCommand(Integer.parseInt(fullCommandArr[1]));
+        case exit:
+            return new ExitCommand();
+        default:
+            return new InvalidCommand();
```
Neat! An example of good code quality!

pangrwa commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,39 @@
+    public FindCommand(String input) {
+        this.InputToSearch = input;
+    }
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException {
+        ArrayList<Task> foundTasks = taskList.findTasks(this.InputToSearch);
+        if (foundTasks.size() == 0) {
+            ui.showMessage("We could not find any matching tasks.");
+        } else {
+            ui.showMessage("Here are the matching tasks in your list: ");
+            for (Task foundTask : foundTasks) {
+                if (foundTask instanceof Todo) {
+                    ui.showMessage(foundTasks.indexOf(foundTask) + 1 + ". [" + foundTask.getSymbol() + "] "
+                            + "[" + foundTask.getStatusIcon() + "] " + foundTask.getDescription());
+                } else {
+                    ui.showMessage(foundTasks.indexOf(foundTask) + 1 + ". [" + foundTask.getSymbol() + "] "
+                                  + "[" + foundTask.getStatusIcon() + "] " + foundTask.getDescription()
+                                  + " (" + foundTask.getDuedateString() + ")");
+                }
+            }
+        }
+    }
```
Seems like an arrow-head style code, maybe perhaps the method is too long and too deeply nested. Maybe try abstracting some of the code into another method?

pangrwa commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,31 @@
+
+public class ListCommand extends Command {
+
+    public void execute(TaskList taskList, Ui ui, Storage storage) {
+        if (taskList.getNumTasks() == 0) {
+            ui.showMessage("You have no tasks due!");
+        } else {
+            ui.showMessage("Here are the tasks you have due!");
+            for (Task cur : taskList.getTasks()) {
+                if (cur instanceof Todo) {
+                    ui.showMessage(taskList.getTasks().indexOf(cur) + 1 + ". [" + cur.getSymbol() + "] "
+                                    + "[" + cur.getStatusIcon() + "] " + cur.getDescription());
+                } else {
+                    ui.showMessage(taskList.getTasks().indexOf(cur) + 1 + ". [" + cur.getSymbol() + "] "
+                                        + "[" + cur.getStatusIcon() + "] " + cur.getDescription()
+                                        + " (" + cur.getDuedateString() + ")");
+                }
+            }
+        }
+    }
```
This method is also quite long and deeply nested! Try abstracting the code out!

pangrwa commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+     * @param input Description of task including its due date.
+     * @param isDone Whether the task is marked or unmarked.
+     */
+    public Deadline(String input, boolean isDone) {
+        super(input, isDone);
+        this.symbol = 'D';
+        String[] temp = input.split(",");
+        this.description = temp[0];
+        LocalDate inputFormatter = LocalDate.parse(temp[1]);
+        this.duedate = inputFormatter;
+        this.duedateString = inputFormatter.format(DateTimeFormatter.ofLocalizedDate(FormatStyle.LONG));
```
Seems like it's not following SLAP, but I personally don't think it's a big issue here.

pangrwa commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+     * @param input Description and duedate of the task, including its header "deadline".
+     * @throws DukeException If description is empty, or duedate keyed in does not fit specified format.
+     */
+    public Deadline(String input) throws DukeException {
+        super(input);
+        this.symbol = 'D';
+        String[] inputArr = input.split(" ", 2); //split 'deadline' from task input
+        if (inputArr.length == 1 || inputArr[1].isBlank()) {
+            throw new DukeException("Sorry, the description of a deadline cannot be empty!");
+        }
+        String[] descriptionArr = inputArr[1].split("/"); //split task from due date
+        if (descriptionArr.length == 1 || descriptionArr[0].isEmpty()) {
+            throw new DukeException("Please include a description or a deadline in the following format: '/yyyy-MM-dd'");
+        }
+        this.description = descriptionArr[0];
+        LocalDate inputFormatter = Parser.parseDate(descriptionArr[1]);
+        this.duedate = inputFormatter;
+        this.duedateString = inputFormatter.format(DateTimeFormatter.ofLocalizedDate(FormatStyle.LONG));
+    }
```
Doesn't seem like it's following SLAP too, but I personally don't think it's a big issue too.

pangrwa reviewed (commented):
Overall, your code is great. Generally follows the SLAP principle except for some of the tasks files. Some methods have arrow-head style code too!

Bisceto committed:
Added a GUI via JavaFX for Duke

Bisceto committed:
No assertions are utilised in the codebase.

Assertions should be used liberally for testing to spot potential bugs quickly

Let's add assertions,
* To Duke.java to ensure that file text is loaded for loading and savings tasks.
* To AddCommand.java to ensure that task is successfully in task list
* To DeleteCommand.java to ensure that task is successfully removed from task list
* To TaskList.java to ensure that commands are successfully marked and unmarked.

Bisceto committed:
Merge pull request #2 from Bisceto/branch-A-Assertions

No assertions are utilised in the codebase.

Bisceto committed:
Some Command methods were unnecessarily long and cluttered.

Let's rectify these comments by updating their methods.
Let's remove the QueryType class to fit into the Query class

Bisceto committed:
Merge pull request #3 from Bisceto/branch-A-CodeQuality

Some Command methods were unnecessarily long and cluttered.

Bisceto committed:
Find command: make matching case insensitive and matches to dates

Find functionality is case sensitive, and only matches to tasks description.

A case insensitive find is more user-friendly because users cannot be expected to remember exact case of keywords.
A find feature that matches dates is more user-friendly for users to see what events and deadlines they have on that day.

Let's,
* update the search algorithm to use case-insensitive matching
* update the search algorithm to match dates in the form <yyyy-MM-dd> or <MMM dd yyyy>

Bisceto committed:
Merge pull request #4 from Bisceto/branch-B-BetterSearch

Find command: make matching case insensitive and matches to dates

Bisceto committed:
Find Command: Bug fixes with regards to matching dates

Find Command contains bugs when keying in dates of different formats.

Let's,
* Update the find command so that it can take in dates <yyyy-MM-dd> or <MMM-dd-yyyy>

Bisceto committed:
Update README.md

Added user guide for Duke chatbot, mainly the list of features that can be used as well as their formats.

Bisceto committed:
Added screenshot of application to showcase its UI

Bisceto committed:
Changed Screenshot of UI to a png file

Bisceto committed:
no message

Bisceto committed:
Update README.md

Bisceto committed:
Mark, List, Delete commands not handling exceptions when index given is out of bounds.

Let's,
Fix the error handling for above methods to accommodate out of bounds input.
